### DISCUSSION TITLE
try-catch expression / try, catch and finally with expression bodies

### DISCUSSION THREAD

**Discussion by Mafii** [dotnet/csharplang#5466] [Language Ideas]
When writing try-catches, a specific pattern used by developers often looks like this:

```cs
// could be anything really
string? result = default;
try 
{
  result = GetValueOrThrow();
}
catch
{
  // Report Error
  throw; // or return default value
}
finally
{
  // this occurs less often in code, but still sometimes
}
```

This specific case takes a minumum of 4 lines of code, not matter how consice your code is, if you adhere to typical C# coding guidelines (brackets on newlines always).

So I would like to suggest the following syntax, that should (to my knowledge) not create any conflicts with already existing or propsed syntax:

```cs
try => GetValueOrThrow()
catch => default(T); 

// additional example
try => GetValueOrThrow()
catch(Exception ex) => { Log(ex); throw; }
finally => resource.Dispose();
```

One could argue that allowing statement bodies for that expression is unreasonable, or that the arrow could just be left out. 

I think however, that just like with expression switches, the language would benefit from a new concept, that does not share syntax and functionality (other than some keywords) with the original feature (try catch here, statement switch in the case of statement expressions). Also, that syntax is very well known and behavies intuitively to all developers that use lambdas and LINQ extensively.

---

Here's how the usage would look in a method:

```cs
public string? GetValueOrNull(Func<T> maybeThrowAction)
  => try => maybeThrowAction()
     catch => null;
```

It would also allow people to use low profile try-catches in longer LINQ-chains, without having to create a method that wraps the try catch. That is a lot of convenience:

```cs
public int GetUserInputCountsOfNumber5(string userInput)
    => string.Split(' ')
      .Select(userNumber =>
        try => int.Parse // yes I know there is TryParse but this is just an example
        catch => 0)
      .Where(// do whatever)
      .Count(number => number == 5);
```

> Does this add value to the language?

I think yes.

> Isn't it just a gimmick?

Yes and no. I have had situations where I wrote some (statement) try catch, and asked a work colleague _"try-catch expression WHEN?"_ It would have made my code shorter, with less fluff, no variable up front so I can use it after the scope, so a try-catch expression returning a value would've been more direct on what my code was doing.

> "But that's not the C# I know?"

C# has been moving to a more functional place in the last few years and versions. Be it immutable records, the with syntax, LINQ, expression switches and many more. Sometimes it feels very out of place writing statement switches or try catches in the "old" style, that is very verbose with those opening scopes. For those people writing short functions, using expression bodies, expression chains, trying to be consice and immutable, a _expression try_ would be a nice addition to their toolbox, as it fits in neatly and itegrates very well with all the newer concepts. The old try catch doesn't have to be abandoned, but it might not be used as frequently by develpers anymore: Such is also the fate of `goto`, `for` loops, `switch` statements, and in my case, even `for each` is rarely seen.

I would love to hear feedback and am open to providing more examples. It would spark great joy to find out, sooner or later, that this or a similar feature would find it's way into my daily drive language!

---

Optionally, if-expressions that act similar would be nice too, but that's out of scope here. They would have similar properties, so I thought I mention them here. Maybe they can come together and move C# to a more functional place in those cases where it makes code easier to read.

**Comment by svick** [comment]
This, or something like this with a different syntax, was previously discussed in e.g. https://github.com/dotnet/csharplang/discussions/4676 and https://github.com/dotnet/csharplang/issues/1567.

**Reply by Mafii** [reply]
#4676 proposes poor syntax (imo) and modification of the existing try catch, which differentiates it from my proposal.

#1567 is a possibility, but it is closed, and all marked duplicates have poorly integrating syntax that seems alien to C# and it is quite old, C# has changed.

So in my opinion, this should be reconsidered.

**Comment by gulshan** [comment]
Some of my exploration here #2734
